File: ws01_plumbing/src/py02_service/py02_service/demo01_server.py

import rclpy
from rclpy.node import Node

from base_interfaces_demo.srv import AddInts

'''
1. 创建服务端
2. 编写回调函数

'''


class AddIntsServer(Node):
    def __init__(self):
        super().__init__("addints_server_py")
        self.get_logger().info("创建服务端 Python")

        self.server = self.create_service(AddInts, "add_ints", self.add)

    def add(self, req, res):
        res.sum = req.num1 + req.num2
        self.get_logger().info("%d + %d = %d" % (req.num1, req.num2, res.sum))
        return res
        



def main():
    rclpy.init()
    rclpy.spin(AddIntsServer())
    rclpy.shutdown()





if __name__ == '__main__':
    main()

File: ws01_plumbing/src/cpp02_service/src/demo01_server.cpp

#include "rclcpp/rclcpp.hpp"

#include "base_interfaces_demo/srv/add_ints.hpp"

using base_interfaces_demo::srv::AddInts;
using std::placeholders::_1;
using std::placeholders::_2;
class AddIntsServer: public rclcpp::Node{
public:
    AddIntsServer():Node("addints_server_cpp"){
        RCLCPP_INFO(this->get_logger(), "创建服务端节点 CPP");

        server_ = this->create_service<AddInts>("add_ints", std::bind(&AddIntsServer::add, this, _1, _2));
    }

private:
    void add(const AddInts::Request::SharedPtr req, const AddInts::Response::SharedPtr res)
    {
        res->sum = req->num1 + req->num2;
        RCLCPP_INFO(this->get_logger(), "%d + %d = %d", req->num1, req->num2, res->sum);
    }
  rclcpp::Service<AddInts>::SharedPtr server_;
};


int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<AddIntsServer>());
    rclcpp::shutdown();
    return 0;
}

